src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -49,79 +49,68 @@ Shard::Shard(DatabaseShard const& db, std::uint32_t index,
 }
 
 bool
-Shard::open(Section config, Scheduler& scheduler)
+Shard::open(Section config, Scheduler& scheduler, nudb::context& ctx)
 {
     assert(!backend_);
     using namespace boost::filesystem;
+    using namespace boost::beast::detail;
 
-    bool dirPreexist;
-    bool dirEmpty;
-    try
+    std::string const type (get<std::string>(config, "type", "nudb"));
+    auto factory {Manager::instance().find(type)};
+    if (!factory)
     {
-        if (!exists(dir_))
-        {
-            dirPreexist = false;
-            dirEmpty = true;
-        }
-        else if (is_directory(dir_))
-        {
-            dirPreexist = true;
-            dirEmpty = is_empty(dir_);
-        }
-        else
-        {
-             JLOG(j_.error()) <<
-                "path exists as file: " << dir_.string();
-            return false;
-        }
+        JLOG(j_.error()) <<
+            "shard " << index_ <<
+            ": failed to create shard store type " << type;
+        return false;
     }
-    catch (std::exception const& e)
+
+    boost::system::error_code ec;
+    auto const preexist {exists(dir_, ec)};
+    if (ec)
     {
         JLOG(j_.error()) <<
-            "shard " + std::to_string(index_) + " exception: " + e.what();
+            "shard " << index_ << ": " << ec.message();
         return false;
     }
 
+    config.set("path", dir_.string());
+    backend_ = factory->createInstance(
+        NodeObject::keyBytes, config, scheduler, ctx, j_);
+
     auto fail = [&](std::string msg)
     {
-        JLOG(j_.error()) <<
-            "shard " << std::to_string(index_) << " error: " << msg;
-
-        if (!dirPreexist)
-            removeAll(dir_, j_);
-        else if (dirEmpty)
+        if (!msg.empty())
         {
-            for (auto const& p : recursive_directory_iterator(dir_))
-                removeAll(p.path(), j_);
+            JLOG(j_.error()) <<
+                "shard " << index_ << ": " << msg;
         }
+        if (!preexist)
+            removeAll(dir_, j_);
         return false;
     };
 
-    config.set("path", dir_.string());
     try
     {
-        backend_ = Manager::instance().make_Backend(
-            config, scheduler, j_);
-        backend_->open(!dirPreexist || dirEmpty);
+        backend_->open(!preexist);
 
+        // If the file descriptor limit is zero, then
+        // the backend does not use permanent storage
         if (backend_->fdlimit() == 0)
```

Comment:
This seems like it should be a member function inside of the `backend_` interface...
`bool isPermanentStorage() const`... which internally might be `return fdlimit() == 0;` but then you don't need a comment explaining it here...

Reply:
Agreed, will implement. 👍

Reply:
👍